ycmd/tests/subcommands_test.py

```diff
@@ -88,6 +88,404 @@ def RunCompleterCommand_GoTo_Clang_ZeroBasedLineAndColumn_test():
+    { 'request': [15,13], 'response': [15,13] }, # sic
+    # main -> definition of main (not declaration)
+    { 'request': [22,7 ], 'response': [22,5] }, # sic
+  ]
+
+  RunSubcommandGoToBatch(contents, tests, ['GoToDefinition']);
+
+  # GoTo - identical to GoToDeclaration
+  #
+  # the semantics of this seem the wrong way round to me. GoToDefinition should
+  # go to where a method is implemented, not where it is declared.
+  #
+  tests = [
+    # Local::x -> declaration of x
+    { 'request': [24,25], 'response': [5 ,9 ] },
+    # Local::in_line -> declaration of Local::inline
+    { 'request': [25,29], 'response': [7 ,10] },
+    # Local -> declaration of Local
+    { 'request': [25,19], 'response': [3 ,11] },
+    # sic: Local::out_of_line -> definition of Local::out_of_line
+    { 'request': [26,30], 'response': [15,13 ] }, # sic
+    # sic: GoToDeclaration on definition of out_of_line moves to itself
+    { 'request': [15,13], 'response': [15,13] }, # sic
+    # main -> definition of main (not declaration)
+    { 'request': [22,7 ], 'response': [22,5] }, # sic
+  ]
+
+  RunSubcommandGoToBatch(contents, tests, ['GoTo']);
+
+  # GoToImprecise - identical to GoToDeclaration
+  #
+  # the semantics of this seem the wrong way round to me. GoToDefinition should
+  # go to where a method is implemented, not where it is declared.
+  #
+  tests = [
+    # Local::x -> declaration of x
+    { 'request': [24,25], 'response': [5 ,9 ] },
+    # Local::in_line -> declaration of Local::inline
+    { 'request': [25,29], 'response': [7 ,10] },
+    # Local -> declaration of Local
+    { 'request': [25,19], 'response': [3 ,11] },
+    # sic: Local::out_of_line -> definition of Local::out_of_line
+    { 'request': [26,30], 'response': [15,13 ] }, # sic
+    # sic: GoToDeclaration on definition of out_of_line moves to itself
+    { 'request': [15,13], 'response': [15,13] }, # sic
+    # main -> definition of main (not declaration)
+    { 'request': [22,7 ], 'response': [22,5] }, # sic
+  ]
+
+  RunSubcommandGoToBatch(contents, tests, ['GoToImprecise']);
+
+def RunSubcommandMessageBatch(contents, tests, command):
+  app = TestApp( handlers.app )
+
+  common_args = {
+    'completer_target'  : 'filetype_default',
+    'command_arguments' : command,
+    'compilation_flags' : ['-x', 
+                           'c++', 
+                           '-std=c++11'],
+    'line_num'          : 10,
+    'column_num'        : 3,
+    'contents'          : contents,
+    'filetype'          : 'cpp'
+  }
+
+  for test in tests:
+    args = test[0]
+    expected = test[1];
+
+    # helps when the test fails to see which one failed
+    print test
+
+    request = common_args
+    request.update( args )
+
+    request_data = BuildRequest( **request )
+
+    eq_( {'message': expected},
+         app.post_json( '/run_completer_command', request_data ).json )
+
+
+@with_setup( Setup )
+def RunCompleterCommand_GetType_Clang_test():
+  contents = """
```

Comment:
Yes. This is much better. I just copied the only other clang completer test in there. The out-of-file tests are about a million times better because you can actually look at the line/column numbers in vim and test it manually. Indeed, that's how I got the test data in the first place.